ls1intum/Artemis
# Generalize Text, Modeling and File-Upload submission process

### Checklist

### Motivation and Context


A lot of code duplication was introduced when adding new exercise type - File Upload. Resolves the issue #819. 
Fixes #1002 .

### Description

- Created `GenericSubmissionResource` to define properties that are used through modeling, text and file-upload submissions. Additionally it has `checkExerciseValidity` used by text and file upload submission resources. 
- Created `GenericSubmissionRepository` which will be extended by concrete repositories. We have some shared queries that can be reused by passing `entityName`
- Moved some of the methods to `SubmissionService`
- `SubmissionRepository` now has `countByCourseIdSubmittedBeforeDueDate` and `countByExerciseIdSubmittedBeforeDueDate` which now additionally check for modeling, text and file upload submission types. Theses allows to reduce similar code in course and exercise resources and make only one request to the database instead of 3.
- Completely removed `GET` Submission endpoint, because it was never used by the client. This endpoint didn't hide any details, so tutor can check which student submitted text exercise. This is also the only client change I had to do.
- Removed `participation-without-assessment` for text exercise from `ParticipationResource`. Instead created a mechanism to handle lock submission in the `TextSubmissionResource`. This should fix the issue with new assessments not loaded and removed code specific for `TextSubmission` from `ParticipationResource`.
- Additional tests for `TextSubmission` and `FileUploadSubmission`.

There are still some lines of code that looks like they can be reused, but due to not having the type of Generic on run time I couldn't generalize them. If you have any further ideas, be sure to leave them in the comments.
Note:
These changes are only related to the submission process, changes to the assessment process will be done in a separate PR.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Overview
3. Select any course with open exercises
4. Submit File Upload/Modeling/Text exercise.
5. Go to Course Administration 
6. Open Tutor Course Dashboard
7. Check that you can see the submission and assessment works properly.
